v7/static_tagcloud/static_tagcloud/engine.py

```diff
@@ -0,0 +1,216 @@
+    ``create_tag_cloud_html`` to create the tag cloud's CSS and HTML code,
+    respectively.
+
+    The argument ``max_number_of_levels`` determines the maximal number of
+    levels. The argument ``max_tags`` ensures that no more than the given
+    number of tags appear in the cloud. The argument
+    ``minimal_number_of_appearances`` will prune all elements with a lower
+    ``count`` from the input ``tags`` list.
+    """
+    # Group tags by weights
+    weights = dict()
+    for i, (name, count, url) in enumerate(tags):
+        if count < minimal_number_of_appearances:
+            continue
+        if count not in weights:
+            weights[count] = []
+        weights[count].append((name, url, i))
+
+    weighted_tags = sorted(weights.items(), key=lambda x: -x[0])
+
+    # Ensure max_tags
+    if max_tags > 0:
+        # Make sure there are at most max_tags tags, by dropping tasks with lowest weight
+        count = 0
+        for weight, tags in weighted_tags:
+            if count < max_tags:
+                if count + len(tags) > max_tags:
+                    tags.sort(key=lambda tag: hashlib.md5(tag[0].encode('utf-8')).hexdigest())
+                    # ... deterministic random sort ...
+                    del tags[max_tags - count:]
+            else:
+                tags.clear()
+            count += len(tags)
+
+        while len(weighted_tags) > 0 and len(weighted_tags[-1][1]) == 0:
+            del weighted_tags[-1]
+
+    # Get list of weights
+    weights = [weight for weight, tags in weighted_tags]
+
+    # Cluster weights into levels
+    clusters = []
+    for left_count in range(max_number_of_levels, 1, -1):
+        if len(weights) == 0:
+            break
+        min = weights[-1]
+        max = weights[0]
+        split = min + (max - min) * (left_count - 1) / float(left_count)
+        cluster = []
+        while len(weights) > 0 and weights[0] > split:
+            cluster.append(weights[0])
+            del weights[0]
+        clusters.append(cluster)
+    clusters.append(weights)
+    while [] in clusters:
+        clusters.remove([])
+
+    # Determine level weights
+    level_weights = [float(i) / len(clusters) for i in range(len(clusters), 0, -1)]
+
+    # Map tag weights to level weights
+    tag_weight_to_level_weight = dict()
+    for index, (level_weight, cluster) in enumerate(zip(level_weights, clusters)):
+        for tag_weight in cluster:
+            tag_weight_to_level_weight[tag_weight] = (level_weight, index)
+
+    # Compute tags
+    tags = []
+    for tag_weight, list in weighted_tags:
+        if tag_weight in tag_weight_to_level_weight:
+            level_weight, level_index = tag_weight_to_level_weight[tag_weight]
+            for name, url, index in list:
+                tags.append((index, name, level_weight, level_index, url))
+    tags.sort()
+    return tags, level_weights
+
+
+def _get_hex_color(color):
+    def myHex(i, digits):
```

Comment:
String formatting would work better for that, and `str.join` can take a generator expression:

```py
def _get_hex_color(color):
    return '#' + ''.join('{:02x}'.format(int(round(c * 255))) for c in color)
```